Maintenance fixes for dependency updates

Xiretza opened the pull request:
Hi! I know the README says you'll only accept security fixes, but I'm trying my luck with sneaking these maintenance fixes in anyway.

`ed25519` version 1.3.0 changed `Signature::new()` to panic on invalid data, which broke tests and caused deprecation warnings; I fixed that and bumped the dep version.

`build.rs` uses `cc::Tool::cflags_env()`, which was introduced in cc 1.0.3. Specifying a too-low dependency breaks any `-Z minimum-versions` builds downstream, I bumped the dep version there as well.

I also prepared a release commit that bumps the version numbers and updates the changelog, I can of course remove that though if you want to do it yourself.

Xiretza committed:
Fix minimum version of cc dependency

build.rs uses cc::Tool::cflags_env(), which was introduced in cc 1.0.3.

Xiretza committed:
Release 0.2.8

Xiretza committed:
Fix compatibility with new ed25519

ed25519 deprecated Signature::new() in favor of Signature::from_bytes(), and
made Signature::new() panic for invalid signatures.

kpcyrd commented on src/crypto/sign/ed25519.rs:
```diff
@@ -336,7 +336,9 @@ mod test {
             for j in 0..SIGNATUREBYTES {
                 sig[j] ^= 0x20;
-                assert!(!verify_detached(&Signature::new(sig), &m, &pk));
+                if let Ok(sig) = Signature::from_bytes(&sig) {
```
I think the test should fail if `Signature::from_bytes` returns `Err(_)` :)

kpcyrd replied:
Ah I see, this is fine then :)

Xiretza committed:
tests: don't run doctests on non-exported macros

Since Rust 1.62.0, doctests on macros are run even if the macro isn't
marked #[macro_export] [0]. Because the macro can't actually be accessed
externally, we explicitly mark the doctest as ignored so it continues
not to be run.

[0]: https://github.com/rust-lang/rust/pull/96630

kpp reviewed (changes requested):
You can call `set_len` after calling `crypto_box_easy`. It will fix the issue.

Xiretza committed:
fix: harden against UB, silence clippy warnings

Calling Vec::set_len() immediately after Vec::with_capacity() creates
uninitialized values. While this is probably not a problem if
Vec::as_mut_ptr() is the only operation done before the Vec is filled
through FFI, only calling set_len() once the Vec is actually filled is
still cleaner.

kpp commented:
@Xiretza can you fix CI?

Xiretza commented:
Not easily, no. The latest versions of our dependencies simply don't build with rustc 1.36 anymore (because they use `edition="2021"`), in order to fix the build they'd either have to be manually pinned to an older version somehow (without affecting users of the library), or we'd have to mess around with the unstable [`-Z minimal-versions`](https://github.com/rust-lang/cargo/issues/5657). Not sure if either is worth it for a library that's unlikely to see many more contributions anyway.

Xiretza committed:
fix(build.rs): silence clippy::format-push-string

adamreichold commented:
> Not easily, no. The latest versions of our dependencies simply don't build with rustc 1.36 anymore (because they use `edition="2021"`), in order to fix the build they'd either have to be manually pinned to an older version somehow (without affecting users of the library), or we'd have to mess around with the unstable [`-Z minimal-versions`](https://github.com/rust-lang/cargo/issues/5657). Not sure if either is worth it for a library that's unlikely to see many more contributions anyway.

A possibly simpler alternative would be to create an alternative `Cargo.lock` once by locally using `-Z minimal-versions` or `cargo update --precise` and check that into the repository. It could then be used for the MSRV build job without adding a lot of automation besides moving the file into place.

Of course, this could still be considered unreasonable effort consider the status of the crate.

kpp commented:
Or you can bump msrv to satisfy CI. Don’t forget to update the docs in Readme.

Xiretza commented:
Sure, if that's what you want to do - I tried to keep the changes as minimal as possible (only requiring minor version bumps of transitive dependencies), but simply increasing the MSRV from 1.36 to 1.56 would of course also fix this.

kpp commented:
Well, yes. Would you please do it?

Xiretza committed:
Bump MSRV requirement to make CI happy

The latest versions of some dependencies use edition="2021", which
requires rust 1.56. Because CI always uses the latest versions of
dependencies, it too will be bumped to 1.56; because older versions are
then no longer actively tested, the MSRV is updated.

Xiretza commented:
Of course, done! I also fixed the clippy warnings in `build.rs`.

pinkforest commented:
Thanks all for keeping the last light on this crate - it's still used by a lot :purple_heart:

kpp commented:
bors r+

kpp commented:
@dnaq can you publish a new release or give someone else rights to do that?

pinkforest commented:
You could add crate ownership to the team organisation for sodiumoxide - I think @Dylan-DPC-zz has the admin ?
Like f.ex. see cargo-geiger is owned by `github:rust-secure-code:cargo-geiger-admins` which is a distinct team in our WG
That would make anyone belonging into a certain team able to publish and the maintainer bus can move in the github org team
https://crates.io/teams/github:rust-secure-code:cargo-geiger-admins
https://crates.io/crates/cargo-geiger

Dylan-DPC commented:
I dont have permissions on crates.io for it. I am ready if they want to pass it to me

dnaq commented:
I'm looking at doing a release with the changes from this pull request, but there are a lot of unrelated changes as well, mostly related to the order of set_length() calls and so on. Due to me not having the time to do any code review of those changes (even though they look correct with a cursory glance) I'm going to have to back out the changes from this pull request and reopen it.

dnaq commented:
Sorry, decided to archive the project completely. Feel free to fork.
